File: ts-core/src/net/node/nodeId.ts
import { UniqueKey } from "@core/object";
import { Address, AddressClass, AddressType, SerialAddress, UdpAddress, UhfAddress, WebSocketAddress } from "../link";
import { FixedBytesSerdeable, ManualVariantSerdeable, Serdeable, TransformSerdeable } from "@core/serde";
import { match } from "ts-pattern";
import { BufferReader } from "../buffer";

export enum NodeIdType {
    // 特別な意味を持つID
    Broadcast = 0xff,
    Loopback = 0x7f,

    // Address互換のID
    Serial = 0x01,
    UHF = 0x02,
    IPv4 = 0x03,
    WebSocket = 0x04,
}

const bodyBytesSizeOf = (type: NodeIdType): number => {
    switch (type) {
        case NodeIdType.Broadcast:
            return 0;
        case NodeIdType.Loopback:
        case NodeIdType.Serial:
        case NodeIdType.UHF:
        case NodeIdType.IPv4:
        case NodeIdType.WebSocket: {
            const address = Address.table[type as unknown as AddressType];
            return address.bodyBytesSize();
        }
        default:
            throw new Error(`NodeId: invalid type: ${type}`);
    }
};

const addressTypeToNodeIdType = (type: AddressType): NodeIdType => {
    return type as unknown as NodeIdType;
};

export class NodeId implements UniqueKey {
    #type: NodeIdType;
    #body: Uint8Array;

    private constructor(type: NodeIdType, body: Uint8Array) {
        if (body.length !== bodyBytesSizeOf(type)) {
            throw new Error(`NodeId: invalid body length: ${body.length}`);
        }
        this.#type = type;
        this.#body = body;
    }

    static broadcast(): NodeId {
        return new NodeId(NodeIdType.Broadcast, new Uint8Array());
    }

    static loopback(): NodeId {
        return new NodeId(NodeIdType.Loopback, new Uint8Array());
    }

    static fromAddress(address: Address | AddressClass): NodeId {
        if (!(address instanceof Address)) {
            address = new Address(address);
        }
        return new NodeId(addressTypeToNodeIdType(address.type()), address.address.body());
    }

    static readonly serdeable = new ManualVariantSerdeable<NodeId>(
        (nodeId) => nodeId.#type,
        (type) =>
            new TransformSerdeable(
                new FixedBytesSerdeable(bodyBytesSizeOf(type as NodeIdType)),
                (body) => new NodeId(type as NodeIdType, body),
                (nodeId) => nodeId.#body,
            ),
    );

    type(): NodeIdType {
        return this.#type;
    }

    isBroadcast(): boolean {
        return this.#type === NodeIdType.Broadcast;
    }

    isLoopback(): boolean {
        return this.#type === NodeIdType.Loopback;
    }

    equals(other: NodeId): boolean {
        return this.#type === other.#type && this.#body.every((byte, index) => byte === other.#body[index]);
    }

    uniqueKey(): string {
        return `NodeId(${this.#type}, ${this.#body.join(",")})`;
    }

    toString(): string {
        return `NodeId(${this.#type}, [${this.#body.join(", ")}])`;
    }

    display(): string {
        return `${this.#type}(${this.#body.join(", ")})`;
    }

    toJSON(): string {
        return this.display();
    }

    toHumanReadableString(): string {
        if (this.#type === NodeIdType.Broadcast) {
            return "Broadcast";
        }
        if (this.#type === NodeIdType.Loopback) {
            return "Loopback";
        }
        return `${NodeIdType[this.#type]}(${this.toHumanReadableBodyString()})`;
    }

    toHumanReadableBodyString(): string {
        type AddressSerdeable = Serdeable<SerialAddress | UhfAddress | UdpAddress | WebSocketAddress>;
        const serdeable: AddressSerdeable | undefined = match(this.#type)
            .with(NodeIdType.Serial, () => SerialAddress.serdeable)
            .with(NodeIdType.UHF, () => UhfAddress.serdeable)
            .with(NodeIdType.IPv4, () => UdpAddress.serdeable)
            .with(NodeIdType.WebSocket, () => WebSocketAddress.serdeable)
            .otherwise(() => undefined);
        if (serdeable === undefined) {
            return "";
        }

        const address = BufferReader.deserialize(serdeable.deserializer(), this.#body).unwrap();
        return address.toHumanReadableString();
    }
}

File: py-core/py_core/net/node/node_id.py
from dataclasses import dataclass
from enum import Enum
from typing import Self
from py_core.net.link.address import Address, AddressType

from py_core.serde.enum import EnumSerde
from py_core.serde.traits import Deserialize, Serialize


class NodeIdTypeValue(Enum):
    # 特別な意味を持つID
    Broadcast = 0xFF

    # Address互換のID
    Serial = 0x01
    Uhf = 0x02
    Udp = 0x03
    WebSocket = 0x04
    Loopback = 0x7F


@dataclass(frozen=True)
class NodeIdType(EnumSerde[NodeIdTypeValue]):
    @staticmethod
    def enum_type():
        return NodeIdTypeValue

    @classmethod
    def from_address_type(cls, address_type: AddressType) -> Self:
        return cls(NodeIdTypeValue(address_type.value))

    _body_bytes_size = {
        NodeIdTypeValue.Broadcast: 0,
        NodeIdTypeValue.Serial: 1,
        NodeIdTypeValue.Uhf: 1,
        NodeIdTypeValue.Udp: 6,
        NodeIdTypeValue.WebSocket: 6,
        NodeIdTypeValue.Loopback: 0,
    }

    def body_bytes_size(self):
        return self._body_bytes_size[self.value]


@dataclass(frozen=True)
class NodeId(Deserialize, Serialize):
    type: NodeIdType
    body: bytes

    @classmethod
    def deserialize(cls, reader) -> Self:
        id_type = NodeIdType.deserialize(reader)
        body = reader.read_bytes(id_type.body_bytes_size())
        return cls(id_type, body)

    def serialize(self, writer) -> None:
        self.type.serialize(writer)
        writer.write_bytes(self.body)

    def serialized_length(self) -> int:
        return self.type.serialized_length() + len(self.body)

    @classmethod
    def from_address(cls, address: Address) -> Self:
        address_type = NodeIdType.from_address_type(address.type)
        return cls(address_type, bytes(address.get_body_as_bytes()))

    @classmethod
    def broadcast(cls) -> Self:
        return cls(NodeIdType(NodeIdTypeValue.Broadcast), b"")

    def is_broadcast(self) -> bool:
        return self.type.value == NodeIdTypeValue.Broadcast

    @classmethod
    def loopback(cls) -> Self:
        return cls(NodeIdType(NodeIdTypeValue.Loopback), b"")

    def is_loopback(self) -> bool:
        return self.type.value == NodeIdTypeValue.Loopback

    def __str__(self) -> str:
        return f"{self.type}({self.body})"
